Rework forge command invocation (broken EPIPE issue)

antico5 opened the pull request:
This PR changes the handling of the child process spawned when running `forge`, currently only used for formatting. The broken EPIPE error happens when a process tries to write to the stdin of a dead process.

kanej commented on server/src/utils/operatingSystem.ts:
```diff
@@ -17,6 +17,61 @@ export async function runCmd(cmd: string, cwd?: string): Promise<string> {
+        }
+      });
+    } else {
+      reject("Child process has no writable stdin stream");
```
I think we reject with an `Error` in each of these.

kanej commented on server/src/utils/operatingSystem.ts:
```diff
@@ -17,6 +17,61 @@ export async function runCmd(cmd: string, cwd?: string): Promise<string> {
+    // Start the child process
+    const child = exec(command, options, (error, stdout, stderr) => {
+      if (error) {
+        reject(error);
```
I think we need to change this to return immediately. I am not sure how node will handle rejecting then continuing on to resolve.

kanej reviewed (changes requested):
I think we need to clear the timeout on process exit.
I am tempted to say we should rework the timeout Error, so that when it is caught by the try/catch in `onDocumentFormatting` we report a more specific sentry status for timeout: `{ status: "internal_error", result: null }`. This would let us separate out timeouts from anything else.

kanej commented:
As a sidepoint - the codecov here is really painful to review. If it is not pulling its weight we should consider turning it off.
What do you think would be required to bring `execWithInput` under test?

antico5 commented:
@kanej addressed all comments. About codecov, I'd be happy to remove it. We can tell on PR reviews what needs and what doesn't need coverage. For execWithInput, I think it's not viable to unit test. It would involve a lot of stubbing and mocking, and the added value would be almost nonexistent. For a non-stub testing case, we'd have to see what's possible in terms of process manipulation from our test case, and it can be tricky to have it be multi platform.

kanej commented on server/src/services/formatting/forgeFormat.ts:
```diff
@@ -19,34 +22,32 @@ export async function forgeFormat(
+    };
+
+    return [textEdit];
+  } catch (error: any) {
+    if (error.killed) {
+      throw new TimeoutError(FORMAT_TIMEOUT);
+    } else {
+      throw error;
+    }
+  }
```
```suggestion
  } catch (error: unknown) {
    if (error instanceof Error && "killed" in error && error.killed) {
      throw new TimeoutError(FORMAT_TIMEOUT);
    } else {
      throw error;
    }
  }
```
We are trying to avoid `any`, I suspect we will want to add helpers to utils to better support error detection patterns.

antico5 replied:
Good point. I just stumbled upon the "ensureError" logic on hardhat. We could use something like that in the future

kanej commented on server/src/services/formatting/onDocumentFormatting.ts:
```diff
@@ -46,11 +47,17 @@ export function onDocumentFormatting(serverState: ServerState) {
+            return { status: "deadline_exceeded", result: null };
+          } else if (error instanceof ForgeResolveError) {
+            return { status: "failed_precondition", result: null };
+          } else {
+            return { status: "internal_error", result: null };
+          }
```
```suggestion
          } else {
            serverState.logger.error(error);
            return { status: "internal_error", result: null };
          }
```

We effectively short-circuit the error handling of `trackTiming`; it would Sentry log the error. I don't think we want to do that for timeouts or failed preconditions - but what do we think of reporting internal errors?
I don't have a strong feeling either way.

antico5 replied:
As of now, saving a solidity file with syntax errors would cause forge fmt to fail, and that would result in an internal_error. Should I handle the additional case where the program executed but exit code was non zero? and avoid reporting it to sentry ?

kanej replied:
Ah. No lets leave this reporting out completely then. Forge fmt failing on syntax errors is definitely not worth reporting.

kanej commented on server/src/utils/operatingSystem.ts:
```diff
@@ -10,13 +10,66 @@ export async function runCmd(cmd: string, cwd?: string): Promise<string> {
+        reject(err);
+      });
+
+      child.once("spawn", () => {
+        stdin.write(input, (error) => {
+          if (error) {
+            reject(error);
+          }
+          stdin.end();
+        });
+      });
```
```suggestion
      child.once("spawn", () => {
        if (!stdin.writable || child.killed) {
          return reject(new Error("Failed to write to unwritable stdin"));
        }

        stdin.write(input, (error) => {
          if (error) {
            reject(error);
          }
          stdin.end();
        });
      });
```

Do we still need a writable/killed guard, or do we think this is covered by the initial spawn step?

antico5 replied:
I had that guard in place but removed it because I didn't find any way to trigger it. But I guess we can just leave it to be extra sure.

kanej replied:
I would have no problem removing it if we know of no way it could be hit.
